seccomp/src/lib.rs

```diff
@@ -785,12 +774,12 @@ impl SeccompFilterContext {
     /// * `default_action` - Action taken for all syscalls that do not match any rule.
     ///
     pub fn new(
-        rules: HashMap<i64, (i64, Vec<SeccompRule>)>,
+        rules: BTreeMap<i64, Vec<SeccompRule>>,
```

Comment:
As I understand, this will order the rules by the syscall no (in the BPF code). Why not use a plain vector for sending/storing the ruleset list? Then, the order of the elements in the vector would be reflected in the BPF code, and the caller would have control over which ruleset gets checked first.

I.e., if we used a struct to represent the ruleset:

```rust
struct SeccompRuleset {
	syscall_no: i64,
	rules: Vec<SeccompRule>
}
```

and a vector to store all the rulesets, then, we could define the complex set of rules once:

```rust
advanced_ruleset_list = vec![
	allow_syscall_if(most_frequent_syscall, AND_OR_MACROS_AND_STUFF),
	allow_syscall_if(second_most_frequent_syscall, ANOTHER_SET_OF_AND_OR_MACROS),
	...
]
```

and choose how to apply them, based on the seccomp level the user requested:

```rust
ruleset_list = match seccomp_level {
	SECCOMP_LEVEL_ADVANCED => advanced_ruleset_list,
	SECCOMP_LEVEL_BASIC => advanced_ruleset_list.map(|rs| allow_syscall(rs.syscall_no))
}

SeccompFilterContext::new(ruleset_list)
```

Also, this would not require `SeccompFilterContext::allow_all()`, which I find a bit confusing - setting some rules, then resetting them all.

Reply:
Yup now I remember we talked about these issues, then my memory failed me :smiley:

Reply:
I spent the better part of today trying out this approach and I'd like to keep the `map`. The `vec` doesn't give us a good way to get to the rules for a syscall, and it's tedious (and overcomplicates the code) to add functionality for modifying in place, **or** inserting if no elements are modifiable (see the `add_rules` function). Moreover, we weren't using the priority approach even when we had support for it; I see no reason to add yet another layer of indirection so that we can have a functionality we've never used.

If you feel strongly about this, I will add another `struct` that wraps a `vec`, so that we can keep the mutation benefits of a `map` while also preserving order (although I'd rather not).

Reply:
I see what you mean, but I found that `add_rules` was only used within tests. Also, I don't think there's anything preventing us from having multiple rulesets for the same syscall.

Anyway, here's the idea I was talking about, in code form: [seccomp-vec](https://github.com/dhrgit/firecracker/blob/seccomp-vec/seccomp/src/lib.rs)

I think there's also some confusion caused by how we named things in that crate (e.g. rules, conditions, etc), since they don't accurately depict the logical/functional flow. I think there is still room for simplification, but this PR may already become too complex itself. Anyway, let me know what you think.

Reply:
Please open issues for any further simplifications you have in mind. This PR set out to fix one problem, and has already become hard to follow and to review because of refactoring iterations. At this point, more changes would only make life difficult for everyone.

Reply:
Okay then - let's merge this as it is now. I'll open an issue for whenever we find time to take this on.